src/Feldspar/Onnx/onnxToFeld.hs

```diff
@@ -29,67 +29,102 @@
+import qualified Onnx.AttributeProto as A
+import qualified Onnx.TypeProto as T
+import qualified Onnx.TypeProto.Value as TV
+import qualified Onnx.TypeProto.Tensor as TT
+import qualified Onnx.TensorProto as TP
+import qualified Onnx.TensorProto.DataType as TD
+import qualified Onnx.TensorShapeProto as TSP
+import qualified Onnx.TensorShapeProto.Dimension as TSP
+import qualified Onnx.TensorShapeProto.Dimension.Value as TSP
+
+import Text.ProtocolBuffers (messageGet)
 import Text.ProtocolBuffers.Header as H (Utf8(..))
 
+import System.FilePath(takeBaseName)
+
 import qualified Data.ByteString.Lazy as L
 import qualified Data.ByteString.Lazy.UTF8 as U
 
-import Data.Either (either)
 import qualified Data.Foldable as D (toList)
+import qualified Data.Set as S
 import Data.List (intercalate)
-import Data.Maybe (maybe, fromJust, fromMaybe)
+import Data.Maybe (fromJust, fromMaybe)
 import System.Environment (getArgs)
+import System.IO (IOMode(WriteMode), openFile, hClose, hPutStr)
 
 main :: IO ()
 main = do args <- getArgs
           let [modelFileName] = take 1 args -- First argument is file name
-          -- putStrLn modelFileName
+              modelBaseName = takeBaseName modelFileName
+              dataFileName = modelBaseName ++ ".data"
+              progFileName = modelBaseName ++ ".hs"
           modelBinary <- L.readFile modelFileName
-          -- putStrLn "Read it!"
           let model = either (\s -> error $ "Error: " ++ s) fst $ messageGet modelBinary
-              gr = fromMaybe (error "No graph in model") $ graph model
-          -- putStrLn $ maybe (error "Graph has no name") toStr $ G.name gr
-          let graphNodes = D.toList $ node gr
-          -- putStrLn $ show $ length graphNodes
-          -- putStrLn "module Inits where"
-          -- putStrLn "import Feldspar"
-          -- putStr   "-- Num graph initializers "
-          -- putStrLn $ show $ length $ D.toList $ G.initializer gr
-          -- putStrLn "-- Graph initializers"
-          -- putStrLn "typedef float Float;"
-          -- putStrLn $ unlines $ map showGIc $ take 1000 $ D.toList $ G.initializer gr
-          putStrLn "Graph inputs"
-          putStrLn $ unlines $ map showVI $ D.toList $ G.input gr
-          putStrLn "Graph nodes"
-          putStr $ unlines $ map showNode graphNodes
-
-showNode :: NodeProto -> String
-showNode n = "  " ++ outs ++ " = " ++ op ++ " " ++ ins ++ "\n" ++ attrs
-  where ins = unwords $ map toStr $ D.toList $ N.input n
-        -- nameStr = maybe (error "No name string") toStr $ N.name n
-        op = "onnx" ++ toStr (fromJust $ N.op_type n)
-        outs = unwords $ map toStr $ D.toList $ N.output n
-        attrs = unlines $ map showAttribute $ D.toList $ N.attribute n
-
-showAttribute :: AttributeProto -> String
-showAttribute a = "  -- " ++ nStr ++ " " ++ val nStr
+              gr = fromMaybe (error "No graph in model") $ O.graph model
+
+          -- Write the weights
+          dfile <- openFile dataFileName WriteMode
```

Comment:
The data file can be 100+M from what I've understood, can we use ByteString to avoid some performance hit from all the IO?

Reply:
I was planning to improve performance as a separate commit. Correct first, then fast! I will try ByteString.